backend/projects/rdf_models.py

```diff
@@ -1,9 +1,45 @@
+from rdflib import Graph, IdentifiedNode
+from typing import Iterable
+from django.conf import settings
+
 from triplestore.constants import AS, EDPOPCOL
 from triplestore.rdf_model import RDFModel
-from triplestore.rdf_field import RDFUniquePropertyField
+from triplestore.rdf_field import RDFUniquePropertyField, RDFQuadField
+from triplestore.utils import Quads
+
+
+class CollectionsField(RDFQuadField):
+    def get(self, instance: RDFModel):
+        return [
+            s
+            for (s, p, o, g) in self._stored_quads(instance)
+        ]
+
+    def _stored_quads(self, instance: RDFModel) -> Quads:
+        store = settings.RDFLIB_STORE
+        results = store.query(f'''
+        SELECT ?col WHERE {{                              
+                ?col a edpopcol:Collection ;
+                as:context <{instance.uri}> .
+        }}
+        ''', initNs={'as': AS, 'edpopcol': EDPOPCOL})
+
+        return [
+            (result, AS.context, instance.uri, Graph(store, result))
+            for (result, ) in results
+        ]
```

Comment:
I am not sure, but it looks to me as if the query in `_stored_quads` gets the information that `get` is looking for. However, `_stored_quads` decorates this information with additional data, which `get` then peels off again. Wouldn't it be more straightforward and efficient to put the query in `get` and then call `get` from `_stored_quads`?

Reply:
I get the confusion. The reason is that `_stored_quads` is also used to update the triplestore. When you call `save()` on the model, the field will compare `_stored_quads()` with `_quads_to_store()` to update the graph. `_stored_quads()` is also used when you call `delete()`.

So `_stored_quads()` returns the representation that is saved in the triplestore, which is needed to make updates. `get()` translates the graph representation to one that is more convenient in the model.

This is implemented on the parent class (`RDFQuadField`), but it looks convoluted in isolation.

Reply:
Sure, but you are in full control here because you are overriding both methods. What stops you from writing this?

```suggestion
    def get(self, instance: RDFModel):
        store = settings.RDFLIB_STORE
        results = store.query(f'''
        SELECT ?col WHERE {{                              
                ?col a edpopcol:Collection ;
                as:context <{instance.uri}> .
        }}
        ''', initNs={'as': AS, 'edpopcol': EDPOPCOL})
        return [
            s
            for (s, ) in results
        ]

    def _stored_quads(self, instance: RDFModel) -> Quads:
        store = settings.RDFLIB_STORE
        
        return [
            (result, AS.context, instance.uri, Graph(store, result))
            for result in self.get(instance)
        ]
```

Reply:
Ah right. Nothing, but to explain why it wasn't written like that: in general, `_stored_triples`/`_stored_quads` may contain information that is ignored by the model (and thus not retrievable from the output of `get()`).

For instance, you could write this field to be agnostic about the graph context. In that case, `_stored_quads` should include the graph the `as:context` triple was stored in, because that information is necessary for save/delete operations. But `get()` can filter it out.

So *in general*, it makes sense to write `_stored_triples`/`_stored_quads` first, and then convert that to an internal value in `get()`. In this specific case, though, you're right that the dependency can flow either way.